mcs/class/System.XML/System.Xml.Serialization/XmlTypeMapping.cs

```diff
@@ -646,8 +646,15 @@ public XmlTypeMapElementInfo FindElement (object ob, int index, object memberVal
 			{
 				if (memberValue == null) return null;
 				Type type = memberValue.GetType();
-				foreach (XmlTypeMapElementInfo elem in _itemInfo)
-					if (elem.TypeData.Type == type) return elem;
+				XmlTypeMapElementInfo bestMatch = null;
+				foreach (XmlTypeMapElementInfo elem in _itemInfo) {
+					if (elem.TypeData.Type == type)
+						return elem;
+					if (elem.TypeData.Type.IsAssignableFrom (type) &&
+						(bestMatch == null || bestMatch.TypeData.Type.IsAssignableFrom (type)))
```

Comment:
bestMatch.TypeData.Type.IsAssignableFrom (type) check looks quite wrong as it'll be always true